File: app/lib/core/submission_review_reason.dart
/// Reading `Submission.review_reason` (Issue #122).
///
/// The sidecar records *why* a submission needs a human as one short,
/// machine-readable string rather than prose, so the app can route the
/// reviewer to the right place instead of showing them a sentence
/// (`domain.submission_intake.describe_coverage_issue` and
/// `adapters.submission_intake._describe_flagged_images`). The shape is
/// `<reason>:<value>` clauses joined by `;`, and for the per-question reasons
/// the value is a comma-separated list of question ids:
///
/// ```text
/// answer_area_undefined:q-1;crop_nearly_blank:q-2,q-3
/// ```
///
/// Parsed here, in `core`, rather than inside the review screen: it is a
/// wire format, it has to be read the same way everywhere, and a screen is
/// the one place a format like this cannot be unit tested cheaply.
library;

/// The reason the sidecar records for a crop that came out as good as blank
/// -- the answer area landed on the margin, or the student wrote nothing.
///
/// Kept as a constant next to the parser so the two cannot drift: this
/// string is `domain.submission_intake.NEARLY_BLANK_CROP_REASON`.
const String nearlyBlankCropReason = 'crop_nearly_blank';

/// The question ids [reviewReason] flags with [reason].
///
/// Returns an empty set for `null`, for a reason that is not present, and
/// for anything that does not parse -- an unrecognized string means a newer
/// sidecar is reporting something this build does not know about, and
/// guessing at it would put a warning on the wrong question. Unknown is
/// shown as nothing, never as a claim.
Set<String> questionsFlaggedAs(String? reviewReason, String reason) {
  if (reviewReason == null || reviewReason.isEmpty) {
    return const <String>{};
  }
  final flagged = <String>{};
  for (final clause in reviewReason.split(';')) {
    final separator = clause.indexOf(':');
    if (separator <= 0 || clause.substring(0, separator).trim() != reason) {
      continue;
    }
    for (final id in clause.substring(separator + 1).split(',')) {
      final trimmed = id.trim();
      if (trimmed.isNotEmpty) {
        flagged.add(trimmed);
      }
    }
  }
  return flagged;
}

/// Whether [questionId]'s crop was flagged as nearly blank.
bool hasNearlyBlankCrop(String? reviewReason, String questionId) =>
    questionsFlaggedAs(
      reviewReason,
      nearlyBlankCropReason,
    ).contains(questionId);

File: desktop/src/renderer/core/submission-review-reason.ts
/**
 * Reading `Submission.review_reason` (Issue #122 / INV-158 / INV-159).
 *
 * Wire format: `<reason>:<value>` clauses joined by `;`, where per-question reasons
 * carry a comma-separated list of question ids:
 * `answer_area_undefined:q-1;crop_nearly_blank:q-2,q-3`
 */

export const nearlyBlankCropReason = "crop_nearly_blank";
export const answerAreaUndefinedReason = "answer_area_undefined";

const REASON_LABELS: Readonly<Record<string, string>> = {
  [answerAreaUndefinedReason]: "回答欄が確定できない設問",
  [nearlyBlankCropReason]: "切り出しがほぼ余白の設問",
};

/**
 * Returns the set of question ids flagged with [reason] in [reviewReason].
 *
 * Returns an empty set for null, empty, or unparseable reasons.
 * Unknown reasons are never guessed at (INV-159).
 */
export function questionsFlaggedAs(
  reviewReason: string | null | undefined,
  reason: string,
): Set<string> {
  if (
    reviewReason === null ||
    reviewReason === undefined ||
    reviewReason.trim().length === 0
  ) {
    return new Set<string>();
  }

  const flagged = new Set<string>();
  for (const clause of reviewReason.split(";")) {
    const separator = clause.indexOf(":");
    if (separator <= 0) {
      continue;
    }
    const clauseReason = clause.substring(0, separator).trim();
    if (clauseReason !== reason) {
      continue;
    }
    const idsPart = clause.substring(separator + 1);
    for (const rawId of idsPart.split(",")) {
      const trimmed = rawId.trim();
      if (trimmed.length > 0) {
        flagged.add(trimmed);
      }
    }
  }

  return flagged;
}

/** Whether [questionId]'s crop was flagged as nearly blank. */
export function hasNearlyBlankCrop(
  reviewReason: string | null | undefined,
  questionId: string,
): boolean {
  return questionsFlaggedAs(reviewReason, nearlyBlankCropReason).has(
    questionId,
  );
}

/**
 * Turns machine-readable review_reason flags into a human-readable Japanese summary.
 *
 * Unknown reasons are ignored and never guessed at (INV-159).
 * If all reasons are unknown or no questions are flagged, returns null.
 */
export function describeReviewReason(
  reviewReason: string | null | undefined,
): string | null {
  if (
    reviewReason === null ||
    reviewReason === undefined ||
    reviewReason.trim().length === 0
  ) {
    return null;
  }

  const parts: string[] = [];
  for (const [reason, label] of Object.entries(REASON_LABELS)) {
    const count = questionsFlaggedAs(reviewReason, reason).size;
    if (count > 0) {
      parts.push(`${label}が${count}問`);
    }
  }

  if (parts.length === 0) {
    return null;
  }

  return `${parts.join("、")}あります`;
}
